plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/concurrent/ConcurrentChunkedIndex.java

```diff
@@ -0,0 +1,140 @@
+ *
+ * Contributors:
+ *   NumberFour AG - Initial API and implementation
+ */
+package org.eclipse.n4js.ide.xtext.server.concurrent;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentMap;
+
+import org.eclipse.emf.ecore.resource.ResourceSet;
+import org.eclipse.xtext.resource.impl.ChunkedResourceDescriptions;
+import org.eclipse.xtext.resource.impl.ResourceDescriptionsData;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Maps;
+
+/**
+ * A concurrent map of container handles to {@link ResourceDescriptionsData}, similar to using a {@link ConcurrentMap}
+ * but with some added functionality such as support for {@link IChunkedIndexListener listeners}.
+ * <p>
+ * IMPORTANT: just as {@link ConcurrentMap}, this class does not ensure the thread-safety of the values (i.e. the
+ * {@link ResourceDescriptionsData} instances) being passed in. Thus, client code must either
+ * <ul>
+ * <li>be sure to not modify a {@code ResourceDescriptionsData} instance after passing it to / retrieving it from this
+ * class OR
+ * <li>must create a {@link ResourceDescriptionsData#copy() copy}.
+ * </ul>
+ */
+public class ConcurrentChunkedIndex {
+
+	/** Map of all containers. */
+	protected final Map<String, ResourceDescriptionsData> containerHandle2Data = new HashMap<>();
+	/** Registered listeners. */
+	protected final List<IChunkedIndexListener> listeners = new ArrayList<>();
+
+	/** Listens for changes in a {@link ConcurrentChunkedIndex}. */
+	public interface IChunkedIndexListener {
+		/** Invoked when the index has changed. */
+		public void onIndexChanged(Map<String, ResourceDescriptionsData> changedContainers,
+				Set<String> removedContainers);
+	}
+
+	/** Removes all containers of this index. */
+	public void clear() {
+		ImmutableSet<String> removedContainers;
+		synchronized (this) {
+			removedContainers = ImmutableSet.copyOf(containerHandle2Data.keySet());
+			containerHandle2Data.clear();
+		}
+		notifyListeners(ImmutableMap.of(), removedContainers);
+	}
+
+	/** Returns the data for the container with the given handle or <code>null</code> if no such container. */
+	public synchronized ResourceDescriptionsData getContainer(String containerHandle) {
+		Objects.requireNonNull(containerHandle);
+		return containerHandle2Data.get(containerHandle);
+	}
+
+	/** Sets the data for this container. */
+	public ResourceDescriptionsData setContainer(String containerHandle, ResourceDescriptionsData newData) {
+		Objects.requireNonNull(containerHandle);
+		Objects.requireNonNull(newData);
+		ResourceDescriptionsData oldData;
+		synchronized (this) {
+			oldData = containerHandle2Data.put(containerHandle, newData);
```

Comment:
newData is mutable. Are we sure that no code path will mutate these in an unsafe fashion?

Reply:
For now, I have solved this via the JavaDoc of the containing class `ConcurrentChunkedIndex`. There it is stated that clients are responsible for making sure not to change instances of `ResourceDescriptionsData` before/after passing them into / obtaining them from a `ConcurrentChunkedIndex` or creating a copy.

Last week I have already implemented an alternative introducing an `ImmutableResourceDescriptionsData` and enforcing its use in `ConcurrentChunkedIndex`, but I'm not sure yet if it might lead to far more copies than necessary (also, as a side note, I ran into a problem with fields of `ResourceDescriptionsData` being private and therefore not being able to copy `ImmutableResourceDescriptionsData` efficiently).